core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -226,20 +229,27 @@ class ReplicaFetcherThread(name: String,
     }
   }
 
-  override def buildFetchRequest(partitionMap: Seq[(TopicPartition, PartitionFetchState)]): FetchRequest = {
+  override def buildFetchRequest(partitionMap: Seq[(TopicPartition, PartitionFetchState)]): ResultWithPartitions[FetchRequest] = {
     val requestMap = new util.LinkedHashMap[TopicPartition, JFetchRequest.PartitionData]
+    val partitionsWithError = mutable.Set[TopicPartition]()
 
     partitionMap.foreach { case (topicPartition, partitionFetchState) =>
       // We will not include a replica in the fetch request if it should be throttled.
       if (partitionFetchState.isReadyForFetch && !shouldFollowerThrottle(quota, topicPartition)) {
-        val logStartOffset = replicaMgr.getReplicaOrException(topicPartition).logStartOffset
-        requestMap.put(topicPartition, new JFetchRequest.PartitionData(partitionFetchState.fetchOffset, logStartOffset, fetchSize))
+        try {
+          val logStartOffset = replicaMgr.getReplicaOrException(topicPartition).logStartOffset
+          requestMap.put(topicPartition, new JFetchRequest.PartitionData(partitionFetchState.fetchOffset, logStartOffset, fetchSize))
+        } catch {
+          case e: KafkaStorageException =>
+            partitionsWithError += topicPartition
+            partitionStates.moveToEnd(topicPartition)
```

Comment:
Good point. I should have noticed this.

In addition to the use of `partitionStates.moveToEnd` here, the use of `partitionStates.moveToEnd` in `AbstractFetcherThread.processFetchRequest.updatePartitionsWithError()` also seems unnecessary. This is because partitionStates.moveToEnd() is redundant if we are going to apply `ConsumerFetcherThread.handlePartitionsWithErrors()` and `ReplicaFetcherThread.handlePartitionsWithErrors()` later. Therefore I have updated the patch to remove the use of `partitionStates.moveToEnd` in both cases.